src/protections/FirstMessageIsLink.ts

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Protection } from "./IProtection";
+import { Mjolnir } from "../Mjolnir";
+import { LogLevel, LogService } from "@vector-im/matrix-bot-sdk";
+import { htmlEscape, isTrueJoinEvent, findLink } from "../utils";
+
+export class FirstMessageIsLink extends Protection {
+    private justJoined: { [roomId: string]: string[] } = {};
+    private recentlyBanned: string[] = [];
```

Comment:
These are unbounded caches, which could cause memory issues over time. We should expire values from here as quickly as possible (first message isn't a link or join was too long ago, for example).

Reply:
oh, true, those lines do clean up the cache. There's a lot of joins which happen though, so we do need to expire based on time too.

Making the expiration time configurable would be valuable. I don't think the average instance will stay online for days at a time, but a few hours should be plenty for most types of spam.